Dependencies/xr/Source/ARCore/XR.cpp

```diff
@@ -642,6 +655,158 @@ namespace xr
+    {
+        auto task = arcana::task_from_result<std::exception_ptr>();
+        
+        // Check if permissions are already granted.
+        if (!GetAppContext().checkSelfPermission(PERMISSION_NAME))
+        {
+            // Register for the permission callback request.
+            arcana::task_completion_source<void, std::exception_ptr> permissionTcs;
+            auto permissionTicket
+            {
+                AddRequestPermissionsResultCallback(
+                    [permissionTcs](int32_t requestCode, const std::vector<std::string>& permissionList, const std::vector<int32_t>& results) mutable
+                    {
+                        // Check if this is our permission request ID.
+                        if (requestCode == PERMISSION_REQUEST_ID)
+                        {
+                            // Loop over permission request results, and find the entry corresponding to the camera grant/deny.
+                            for (int i = 0; i < permissionList.size(); i++)
+                            {
+                                if (permissionList[i].find(PERMISSION_NAME) != std::string::npos)
+                                {
+                                    // If the permission is found and granted complete the task.
+                                    if (results[i] == 0 /* PackageManager.PERMISSION_GRANTED */)
+                                    {
+                                        permissionTcs.complete();
+                                        return;
+                                    }
+                                }
+                            }
+
+                            // Did not find permission in the list, or it was explicitly denied.  Complete the task with an error.
+                            std::ostringstream message;
+                            message << "Camera permission not acquired successfully";
+                            permissionTcs.complete(arcana::make_unexpected(make_exception_ptr(std::runtime_error{message.str()})));
+                        }
+                    })
+            };
+
+            // Kick off the permission check request, and set the task for our caller to wait on.
+            GetMainActivity().requestPermissions(PERMISSION_NAME, PERMISSION_REQUEST_ID);
+            task = permissionTcs.as_task().then(runtimeScheduler, arcana::cancellation::none(), [ticket = std::move(permissionTicket)](){});
+        }
+
+        return task;
+    }
+
+    arcana::task<std::shared_ptr<System::Session>, std::exception_ptr> System::Session::CreateAsync(const Babylon::JsRuntimeScheduler& runtimeScheduler, System& system, void* graphicsDevice)
+    {
+        // First perform the ARCore installation check, request install if not yet installed.
+        return CheckAndInstallARCoreAsync(runtimeScheduler).then(runtimeScheduler, arcana::cancellation::none(), [&runtimeScheduler, &system, graphicsDevice]()
```

Comment:
It doesn't matter what thread `CheckAndInstallARCoreAsync` or `CheckCameraPermissionAsync` are called on right? It only matters where the caller continues, since the caller is the one that transitions back to JS by completing a promise. Assuming this is correct, then I would think that this code could just use `inline_scheduler` and the caller could use the JS scheduler to get back on the JS thread before completing the promise.

Reply:
So I don't think at the xr.cpp layer we need to have any knowledge of the `JsRuntimeScheduler`.

Reply:
The OpenGL initialization in the constructor of the session can only occur on the main JS thread as each OpenGL context is only valid on the thread it is created on (which is the JS runtime thread).  So I would either need to go back to breaking this out into two calls for creation & initialization or pass the jsruntime scheduler down.

When I switch over to using the inline_scheduler it successfully runs the final tasks in CheckCameraPermissionsAsync and CheckAndInstallARCore, but does not properly picking up the continuation in the calling method.  We can debug it Monday.

Reply:
Worked with Ryan offline, and got rid of the need to use another scheduler.  There was a bug in our use of ticketed collection using a mutex instead of a recursive_mutex that was causing my other issues, and we moved the initialization code to the first instance of GetNextFrame to get around the OpenGL issues.